diesel/src/macros/associations/has_many.rs

```diff
@@ -0,0 +1,151 @@
+/// Defines a one-to-many association for the parent table. This macro is only requried if you need
+/// to join between the two tables. This macro should be called iwth the name of the child table,
+/// followed by any options, follwed by the entire struct body. The struct *must* be annotated with
+/// `#[table_name(name_of_table)]`. Both the parent and child structs must implement
+/// [`Identifiable`][identifiable].
+///
+/// [identifiable]: prelude/trait.Identifiable.html
+///
+/// # Options
+///
+/// ## foreign_key
+///
+/// Required. The name of the foreing key column for this association.
+///
+/// # Examples
+///
+/// ```no_run
+/// # #[macro_use] extern crate diesel;
+/// # table! { users { id -> Integer, } }
+/// # table! { posts { id -> Integer, user_id -> Integer, } }
+/// pub struct User {
+///     id: i32,
+/// }
+/// # Identifiable! { #[table_name(users)] struct User { id: i32, } }
+///
+/// pub struct Post {
+///     id: i32,
+///     user_id: i32,
+/// }
+/// # Identifiable! { #[table_name(posts)] struct Post { id: i32, user_id: i32, } }
+///
+/// HasMany! {
+///     (posts, foreign_key = user_id)
+///     #[table_name(users)]
+///     struct User {
+///         id: i32,
+///     }
+/// }
+/// # fn main() {}
+/// ```
+///
+/// To avoid copying your struct definition, you can use the
+/// [custom_derive crate][custom_derive].
+///
+/// [custom_derive]: https://crates.io/crates/custom_derive
+///
+/// ```ignore
+/// custom_derive! {
+///     #[derive(HasMany(posts, foreign_key = user_id)]
+///     #[table_name(users)]
+///     struct User {
+///         id: i32,
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! HasMany {
+    // Format arguments
+    (
+        ($child_table_name:ident, foreign_key = $foreign_key_name:ident)
+        $($rest:tt)*
+    ) => {
+        HasMany! {
+            (
+                child_table = $child_table_name::table,
+                foreign_key = $child_table_name::$foreign_key_name,
+            )
+            $($rest)*
+        }
+    };
+
+    // Extract table name from struct
+    (
+        ($($args:tt)*)
+        #[table_name($table_name:ident)]
+        $($rest:tt)*
+    ) => {
+        HasMany! {
+            (
+                parent_table_name = $table_name,
+                $($args)*
+            )
+            $($rest)*
+        }
+    };
+
+    // Strip meta items, pub (if present) and struct from definition
+    (
+        $args:tt
+        $(#[$ignore:meta])*
+        $(pub)* struct $($body:tt)*
+    ) => {
+        HasMany! {
+            $args
+            $($body)*
+        }
+    };
+
+    // Receive parsed fields of normal struct from `__diesel_parse_struct_body`
+    // These patterns must appear above those which start with an ident to compile
+    (
+        (
+            struct_name = $struct_name:ident,
+            parent_table_name = $parent_table_name:ident,
+            child_table = $child_table:path,
+            foreign_key = $foreign_key:path,
```

Comment:
Are these the only fields ever given here? Would it make sense/this more future-compatible if you add a generic token tree match at the end? Like `$($other_headers:tt)*`